### ISSUE TITLE
Adjacency analysis fails on fuel cell model with the NB

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13031]
Please check [SOFCPoliMi.Tests.BenchmarkCammarata](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarata.err). The NB fails with
```
Error: Internal error NBSlice.resolveDependency (single dependency) failed because list of scalar variables(10) {stack.module[$i1].pen.D.pCathode}, does not fit the equation size 100.
Error: Internal error NBSlice.resolveDependency failed for: stack.module[$i1].pen.D.pCathode.
Error: Internal error NBAdjacency.Matrix.upgradeRow failed for:
[FOR-] (100) ($RES_SIM_651)
[----] for $i1 in 1:10 loop
[----]   [ARRY] (10) stack.module[$i1].pen.D.D_O2_i = {(((5.959770780848035e-24 * sqrt(31.25117191894696 + 1/({0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948})[i]) * stack.module[$i1].pen.D.Temp ^ 1.5 * stack.module[$i1].pen.D.ratioPorTor) / stack.module[$i1].pen.D.pCathode) / ({3.054e-10, 3.147e-10, 3.7040000000000003e-10, 3.5785e-10, 3.6125e-10, 3.9550000000000003e-10, 4.2925000000000004e-10, 3.467e-10, 3.6325e-10, 3.5045e-10})[i] ^ 2.0) / stack.module[$i1].pen.D.collIntegral_O2_i[i] for i in 1:10} ($RES_SIM_652)
[----] end for;
```
Can you please check? Thanks!

**Comment by casella** [comment]
Adding @matteodepascali to the loop

**Comment by casella** [comment]
Partially addressed by #13031.

Now [fails](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarata.err) with
```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
```
which may be expected because the experiment is set up with fixed pressure boundary condition that leads to an index-2 system.

**Comment by casella** [comment]
This [commit](https://github.com/looms-polimi/SOFCPoliMi/commit/74ae525ae855ca247e4c8f3e42da7c4bee05729e) added an index-1 test to the library.

Job 8564 should test that. Results will be available [here](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/SOFCPoliMi.html).

**Comment by casella** [comment]
Now [all tests](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/SOFCPoliMi.html), regardless of their index, fail with the same error:
```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
(6|6) Unmatched Variables
***************************
...
(117|654) Unmatched Equations
*******************************
...
```

**Comment by casella** [comment]
You can find smaller examples failing with the same error in the EHPTexamples library, e.g. [EHPTexamples.EV.FirstEV](https://libraries.openmodelica.org/branches/newInst-newBackend/EHPTexamples/files/EHPTexamples_EHPTexamples.EV.FirstEV.err):
```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
(2|2) Unmatched Variables
***************************
        [ALGB] (1) Real mass.a (start = 0.0)
        [ALGB] (1) flow Real gear.flange_b.tau

(2|2) Unmatched Equations
***************************
        [SCAL] (1) dragF.s = mass.s + 0.5 * mass.L ($RES_SIM_46)
        [SCAL] (1) (gear.flange_b.phi - wheel.fixedR.phi0) * wheel.radius = (mass.s - 0.5 * mass.L) - wheel.fixedT.s0 ($RES_SIM_7)
```
I don't know what they have in common with the original problem, but it seems much easier to debug.

**Comment by kabdelhak** [comment]
I fixed two problems, but there seem to be remaining problems.

1. Inlining did not work because incorrect record sizes were used. fixed with #13227
2. Record array lowering was incorrect because the children did not inherit the array dimensions. fixed with #13230

**Comment by kabdelhak** [comment]
After everything is merged the error message is:

```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
(10|16) Unmatched Variables
*****************************
	[ALGB] (10) flow Real[10] stack.module.pen.cathodeHeatPort.Q_flow (nominal = {1e6 for $f1 in 1:10})
	slice: {3}
	[ALGB] (10) Real[10] $FUN_32
	slice: {0, 4, 5, 6, 7}
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}
	[ALGB] (100) Real[10, 10] stack.module.anodeChannel.nOut_i = {stack.module[$module1].anodeChannel.nOut * stack.module[$module1].anodeChannel.fluidOut.Y for $module1 in 1:10} (nominal = {1e-7 for $f1 in 1:10, $f2 in 1:10})
	slice: {90}
	[ALGB] (10) Real[10] stack.module.pen.concLosses.D_H2_Rest (nominal = {1e-5 for $f1 in 1:10})
	slice: {0}
	[ALGB] (1) flow Real stack.cathodeOutletMultiplier.inlet.w (min = 0.0, nominal = 100.0)
	[DER-] (10) Real[10] $DER.stack.module.cathodeChannel.pOut
	slice: {0, 1, 9}
	[ALGB] (1) flow Real stack.anodePin.i (nominal = 0.001)
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2O_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}
	[ALGB] (1) Real stack.enBalance = (stack.module[1].cathodeChannel.entFlowIn + stack.module[1].anodeChannel.entFlowIn) - (stack.pElStack + stack.module[10].cathodeChannel.entFlowOut + stack.module[10].anodeChannel.entFlowOut + Qloss.k * stack.QlossFrac) (nominal = 1e7)

(5|16) Unmatched Equations
****************************
	[SCAL] (1) stack.module[10].cathodeChannel.outlet.p = stack.module[10].cathodeOut.p ($RES_SIM_1738)
	[FOR-] (100) ($RES_SIM_2313)
[----] for {$i1 in 1:10, i in 1:10} loop
[----]   [SCAL] (1) stack.module[$i1].pen.concLosses.collIntegral_O2_i[i] = $FUN_29[i, $i1] ($RES_SIM_2314)
[----] end for;
	slice: {10, 20, 30, 40, 50, 60, 70, 80, 90, 99}
	[FOR-] (10) ($RES_SIM_750)
[----] for $i1 in 1:10 loop
[----]   [SCAL] (1) stack.module[$i1].anodeChannel.pOut = stack.module[$i1].anodeChannel.outlet.p ($RES_SIM_751)
[----] end for;
	slice: {9}
	[FOR-] (100) ($RES_SIM_2309)
[----] for {$i1 in 1:10, i in 1:10} loop
[----]   [SCAL] (1) stack.module[$i1].pen.concLosses.collIntegral_H2_i[i] = $FUN_27[i, $i1] ($RES_SIM_2310)
[----] end for;
	slice: {98, 99}
	[FOR-] (10) ($RES_SIM_982)
[----] for $i1 in 1:10 loop
[----]   [SCAL] (1) stack.module[$i1].anodeChannel.fluidIn.MMmix = 1/(stack.module[$i1].anodeChannel.fluidIn.X * {55.508435061791985, 496.06127348850134, 22.7223667617219, 35.70140770650587, 62.334579609361654, 33.25679835471967, 22.677989333180935, 31.25117191894696, 35.69720205330306, 25.032542304996497}) ($RES_SIM_983)
[----] end for;
	slice: {0, 1}
```

still a singularity issue, but at least everything is balanced now. :) 
Currently looking for where this singularity might come from. Since we currently don't have index reduction, the full singular set is not yet automatically detected, this is just the unmatched sub section of it.

If you have an idea what could be wrong here, or what do look for it would be much appreciated @casella. The system is very large. I think will implement the detection of the singular subsystem to properly analyze this.

**Comment by casella** [comment]
@kabdelhak, [SOFCPoliMi.Tests.BenchmarkCammarata](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarata.err) is known to be index-2. Please try with [SOFCPoliMi.Tests.BenchmarkCammarataIndex1](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarataIndex1.err), if you haven't already done it.

**Comment by casella** [comment]
Barring boundary conditions that impose constraints on differentiated variables (which is the case of the BenchmarkCammarata model), the flat SOFC model has been successfully exported to Pyomo.DAE, where it was used to solve both a simulation problem (zero degrees of freedom, completely constrained) and some optimization problem (free inputs, to be determined to minimize an objective function). Pyomo.DAE handles non-singular index-1 function only, so I am not expecting any structural problem in that model, once you pick the one that has the right boundary conditions.

**Comment by casella** [comment]
@kabdelhak, can you run [EHPTexamples.EV.FirstEV](https://libraries.openmodelica.org/branches/newInst-newBackend/EHPTexamples/files/EHPTexamples_EHPTexamples.EV.FirstEV.err) after your fixes?

**Comment by kabdelhak** [comment]
Unfortunately both models still fail, will have a look at [EHPTexamples.EV.FirstEV](https://libraries.openmodelica.org/branches/newInst-newBackend/EHPTexamples/files/EHPTexamples_EHPTexamples.EV.FirstEV.err) first.

**Comment by kabdelhak** [comment]
Regarding [EHPTexamples.EV.FirstEV](https://libraries.openmodelica.org/branches/newInst-newBackend/EHPTexamples/files/EHPTexamples_EHPTexamples.EV.FirstEV.err): It seems like an index 2 system to me.

There is the equation
```
	[SCAL] (1) dragF.s = mass.s + 0.5 * mass.L ($RES_SIM_46)
```

with
```
(1)       [STAT] (1) Real dragF.s
(3)       [STAT] (1) Real mass.s (start = 0.0, StateSelect = default)
(52)      [PRMT] (1) parameter Real mass.L = 0.0 (fixed = true, start = 0.0)
```
Two states and a parameter. Did you expect this to be index 1 @casella ?

**Comment by casella** [comment]
Of course it is, that's a 1D mechanical system :)

**Comment by casella** [comment]
Maybe also the BenchmarkCammarataIndex1 is actually index-2. I'll check with Dymola ASAP.

**Comment by kabdelhak** [comment]
> Of course it is, that's a 1D mechanical system :)

Ah well then we cannot solve this yet :) I need to implement index reduction. The error you see is just the compiler telling us it cannot do index reduction yet.

**Comment by casella** [comment]
I just ran the BenchmarkCammarataIndex1 model with Dymola, turning on the logging of equation differentiation for index reduction and I saw none. Seems really an index-1 system.

Can I get the same kind of diagnostics with OMC using the old backend?

**Comment by casella** [comment]
@kabdelhak, you suggested me to use the old backend and `-d=bltdump` to check for the system index, but I am getting inconclusive results. In fact, I'm not getting any specific notification regarding index when compiling BenchmarkCammarataIndex1 with that flag.

If I compile this model:
```modelica
model MWE
  Real x, y, z;
equation
  3*der(x) = -z + 1;
  2*der(y) = z;
  x =3* y;
annotation(
    __OpenModelica_commandLineOptions = "-d=bltdump");
end MWE;
```
with the old backend, I indeed get some useful indications about index reduction
```
Index Reduction neccessary! 

Constraint equation to be differentiated: 
x = 3.0 * y 
Differentiated equation: 
der(x) = 3.0 * der(y) 
```
However, if I compile this example
```modelica
model MWE2
  Real x, y, z;
equation
  3*der(x) = -z + 1;
  2*der(y) = z;
  x = y;
annotation(
    __OpenModelica_commandLineOptions = "-d=bltdump");
end MWE2;
```
which is also obviously index-2, with the old bakend. I don't get this indication. From what I understand by adding -`d=optdaedump`, the OB handles this high-index model during the removeSimpleEquations section, going from
```
Variables (3) 
======================================== 
1: z:VARIABLE() MWE type: Real 
2: y:STATE(1)() MWE type: Real 
3: x:STATE(1)() MWE type: Real 
 
 
Equations (3, 3) 
======================================== 
1/1 (1): 3.0 * der(x) = 1.0 - z   [dynamic |0|0|0|0|] 
2/2 (1): 2.0 * der(y) = z   [dynamic |0|0|0|0|] 
3/3 (1): x = y   [dynamic |0|0|0|0|] 
```
to
```
Variables (2) 
======================================== 
1: z:VARIABLE() MWE type: Real 
2: x:STATE(1)() MWE type: Real 
 
Equations (2, 2) 
======================================== 
1/1 (1): 2.0 * der(x) = z   [dynamic |0|0|0|0|] 
2/2 (1): 3.0 * der(x) = 1.0 - z   [dynamic |0|0|0|0|] 

Alias Variables (1) 
======================================== 
1: y:DUMMY_STATE()  = x MWE type: Real 
```
So in this simple case (equality constraint among differentiated variables), a sort of poor man's, alias-based index reduction is performed.

When I try both MWEs with the new backend, both cases are handled by alias elimination (since non-trivial, linear alias elimination is also implemented), while the following example
model MWE3
  Real x, y, z;
equation
  3*der(x) = -z + 1;
  2*der(y) = z;
  x = y*(y+1);
annotation(
    __OpenModelica_commandLineOptions = "-d=bltdump,optdaedump --newBackend");
end MWE3;

which is index-2 with non-trivial state constraint, would need proper index reduction and fails because that is still not provided.

Now, if BenchmarkCammarataIndex1 is indeed index-2, it is because of some trivial `x1 = x2` (or possibly `x1 = a*x2`) state constraint, which is not caught by the OB with `-d=bltdump`. What I suspect is that this trivial alias equation involves arrays and is thus not (yet) processed, as in this MWE
```modelica
model MWE4
  Real x[2], z;
equation
  3*der(x[1]) = -z + 1;
  2*der(x[2]) = z;
  x[1] = x[2];
annotation(
    __OpenModelica_commandLineOptions = "-d=bltdump,optdaedump --newBackend");
end MWE4;
```
which fails with the NB. I can catch this potential problem by running the model with the OB and `-d=optdaedump`, then looking for DUMMY_STATE.

**Comment by kabdelhak** [comment]
Yes you are totally correct, alias elimination handles this case and it might be that in the case of array alias it is not properly handled yet by the new backend. The main problem with `x[1] = x[2]` is that it would need partial array states which is not supported yet (not all of `x` remains a state after reduction).

I have a concept for sliced states but did not have time to work on it yet. Is planned for when index reduction will be implemented.

Furthermore currently we do not do slice replacements in general because one would need to detect if the full array is to be replaced by a full other array. 
```
x[1] = y[1];
x[2] = y[2];
```
Should be possible to replace if either `x` or `y` is of size 2, but currently is not.

**Comment by kabdelhak** [comment]
> Or to the old one?

You could try to run it with `--preOptModules-=removeSimpleEquations` and `-d=bltdump` to see if it has index reduction without the removal, but i am not certain the old backend is stable enough here to not remove simple equations. In general we probably should report state equality removal in both old and new backend.

**Comment by casella** [comment]
Catching up on this topic again.

I just successfully ran `SOFCPoliMi.Tests.BenchmarkCammarataIndex1` with the OB and the flags `--preOptModules-=removeSimpleEquations -d=bltdump` but I got no report of index reduction, as I instead get with this simple MWE:
```modelica
model TestIndexReductionAlias
  Real x1, x2, y;
equation
  der(x1) = 1 - y;
  der(x2) = y;
  x1 = x2;
end TestIndexReductionAlias;
```
that gives:
```
Index Reduction neccessary! 
MSS subsets: 
 3
```

As a further counterproof, I ran `SOFCPoliMi.Tests.BenchmarkCammarata`, which is known to be index-2 because of an equality constraint on a differentiated pressure variable, with `--preOptModules-=removeSimpleEquations -d=bltdump`, and I indeed got the index reduction warning: 
```
Index Reduction neccessary! 
MSS subsets: 
 14150, 557, 641, 14241, 660, 674, 14562 
 13367, 543, 642, 14215, 648, 675, 14563 
```

So, I have now very strong evidence that `SOFCPoliMi.Tests.BenchmarkCammarataIndex1` has indeed index 1:
- the OB doesn't report it as being index 1, when suppressing alias elimination
- when the model is exported via BaseModelica to Pyomo, which only accepts index-1 models, it works fine for both simulation and optimal control

Apparently, something else is going wrong in that model, and is not being high index

**Comment by casella** [comment]
I also observed that there are some similarities in the index-2 and index-1 models.

### Index 2: BenchmarkCammarata
```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
(8|16) Unmatched Variables
****************************
	[ALGB] (10) Real[10] $FUN_32
	slice: {0, 4, 5, 6, 7, 8}
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}
	[ALGB] (100) Real[10, 10] stack.module.anodeChannel.nOut_i = {stack.module[$module1].anodeChannel.nOut * stack.module[$module1].anodeChannel.fluidOut.Y for $module1 in 1:10} (nominal = {1e-7 for $f1 in 1:10, $f2 in 1:10})
	slice: {90}
	[ALGB] (10) Real[10] stack.module.pen.concLosses.D_H2_Rest (nominal = {1e-5 for $f1 in 1:10})
	slice: {0}
	[ALGB] (1) flow Real stack.cathodeOutletMultiplier.inlet.w (min = 0.0, nominal = 100.0)
	[ALGB] (10) Real[10] stack.module.pen.logBoundCHeck = {SOFCPoliMi.Functions.logBound((stack.module[$module1].pen.pH2 * stack.module[$module1].pen.pH2OTPB) / (stack.module[$module1].pen.pH2TPB * stack.module[$module1].pen.pH2O)) + SOFCPoliMi.Functions.logBound(stack.module[$module1].pen.pO2 / stack.module[$module1].pen.pO2TPB) * 0.5 for $module1 in 1:10}
	slice: {0, 1, 3, 9}
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2O_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}
	[ALGB] (1) Real stack.enBalance = (stack.module[1].cathodeChannel.entFlowIn + stack.module[1].anodeChannel.entFlowIn) - (stack.pElStack + stack.module[10].cathodeChannel.entFlowOut + stack.module[10].anodeChannel.entFlowOut + Qloss.k * stack.QlossFrac) (nominal = 1e7)

(5|16) Unmatched Equations
****************************
	[SCAL] (1) stack.module[10].cathodeChannel.outlet.p = stack.module[10].cathodeOut.p ($RES_SIM_1738)
	[FOR-] (100) ($RES_SIM_2313)
[----] for {$i1 in 1:10, i in 1:10} loop
[----]   [SCAL] (1) stack.module[$i1].pen.concLosses.collIntegral_O2_i[i] = $FUN_29[i, $i1] ($RES_SIM_2314)
[----] end for;
	slice: {10, 20, 30, 40, 50, 60, 70, 80, 90, 99}
	[FOR-] (10) ($RES_SIM_750)
[----] for $i1 in 1:10 loop
[----]   [SCAL] (1) stack.module[$i1].anodeChannel.pOut = stack.module[$i1].anodeChannel.outlet.p ($RES_SIM_751)
[----] end for;
	slice: {9}
	[FOR-] (100) ($RES_SIM_2309)
[----] for {$i1 in 1:10, i in 1:10} loop
[----]   [SCAL] (1) stack.module[$i1].pen.concLosses.collIntegral_H2_i[i] = $FUN_27[i, $i1] ($RES_SIM_2310)
[----] end for;
	slice: {98, 99}
	[FOR-] (10) ($RES_SIM_982)
[----] for $i1 in 1:10 loop
[----]   [SCAL] (1) stack.module[$i1].anodeChannel.fluidIn.MMmix = 1/(stack.module[$i1].anodeChannel.fluidIn.X * {55.508435061791985, 496.06127348850134, 22.7223667617219, 35.70140770650587, 62.334579609361654, 33.25679835471967, 22.677989333180935, 31.25117191894696, 35.69720205330306, 25.032542304996497}) ($RES_SIM_983)
[----] end for;
	slice: {0, 1}
```
### Index-1 BenchmarkCammarataIndex1
```
Error: Internal error NBResolveSingularities.noIndexReduction failed.
(5|14) Unmatched Variables
****************************
	[ALGB] (10) Real[10] $FUN_32
	slice: {0, 1, 2, 3, 4, 5, 6, 7, 8}
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}
	[DER-] (10) Real[10] $DER.stack.module.pen.T
	slice: {0, 1}
	[ALGB] (10) Real[10] stack.module.pen.concLosses.D_H2_Rest (nominal = {1e-5 for $f1 in 1:10})
	slice: {0}
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2O_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	slice: {0}

(4|14) Unmatched Equations
****************************
	[SCAL] (1) stack.module[1].anodeChannel.hIn = anodeSource.fluid.h ($RES_SIM_778)
	[FOR-] (100) ($RES_SIM_788)
[----] for $i1 in 1:10 loop
[----]   [ARRY] (10) stack.module[$i1].anodeChannel.fluidOut.X = stack.module[$i1].anodeChannel.Xout ($RES_SIM_789)
[----] end for;
	slice: {90}
	[FOR-] (100) ($RES_$AUX_2449)
[----] for {i in 1:10, $i1 in 1:10} loop
[----]   [SCAL] (1) $FUN_27[i, $i1] = SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.CollisionIntegral(2, i, stack.module[$i1].pen.concLosses.Temp / ({219.78004, 59.7, 107.9511, 73.9898, 94.18822, 113.47815, 118.97424, 79.81222, 65.28844, 74.6325})[i]) ($RES_$AUX_2450)
[----] end for;
	slice: {89, 99}
	[FOR-] (100) ($RES_$AUX_2445)
[----] for {i in 1:10, $i1 in 1:10} loop
[----]   [SCAL] (1) $FUN_29[i, $i1] = SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.CollisionIntegral(8, i, stack.module[$i1].pen.concLosses.Temp / ({293.82132, 79.81222, 144.31854, 98.91608, 125.9191, 151.70758, 159.05524, 106.7, 87.28333, 99.7753})[i]) ($RES_$AUX_2446)
[----] end for;
	slice: {90, 91, 92, 93, 94, 95, 96, 97, 98, 99}
```
In both cases the following variables are involved:

```
	[ALGB] (10) Real[10] $FUN_32
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
	[ALGB] (10) Real[10] stack.module.pen.concLosses.D_H2_Rest (nominal = {1e-5 for $f1 in 1:10})
	[ALGB] (80) Real[10, 8] stack.module.pen.concLosses.D_H2O_i (nominal = {1e-5 for $f1 in 1:8, $f2 in 1:10})
```
though with different slices. I guess this can hardly be a coincidence and probably hints at something going wrong with those variables and their matched equations.

I guess doing the debugging on a 16712-equation model is out of the question 😅 . What we can do is to add some smaller component tests to the library, so we can hopefully pin-point the issue more easily.

**Comment by casella** [comment]
Since in both cases we have some evidence of issue with the concLosses component, we tried to test that component standalone:
```modelica
model Test
  SOFCPoliMi.Components.FuelCell.DiffusionLosses.DiffusivityCoefficients diff(
    pAnode = 30e5+time,
    pCathode = 30e5-time,
    T = 1000+100*time,
    pXa = diff.pAnode*{0.4-0.1*time, 0.1, 0.2, 0.1, 0.15, 0.0, 0.0, 0.05 + 0.1*time, 0.0, 0.0},
    pXc = diff.pCathode*{0.0, 0.0, 0.7-0.1*time, 0.0, 0.0, 0.0, 0.0, 0.3+0.1*time, 0.0, 0.0}
    );
  annotation(
    uses(SOFCPolimi),
    __OpenModelica_commandLineOptions="--newBackend");
end Test;
```
The test works fine with the old backend. With the new one, it goes through the backend but then invalid C code is generated.

```C
Test_functions.c:31:24: error: passing 'void *' to parameter of incompatible type 'base_array_t' (aka 'struct base_array_s')
   31 |   real_array_copy_data(_OMC_LIT0, _table);
      |   ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
./Test_literals.h:7:19: note: expanded from macro '_OMC_LIT0'
    7 | #define _OMC_LIT0 MMC_REFSTRINGLIT(_OMC_LIT_STRUCT0)
      |                   ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/meta/meta_modelica_data.h:188:32: note: expanded from macro 'MMC_REFSTRINGLIT'
  188 | #define MMC_REFSTRINGLIT(NAME) MMC_TAGPTR(&(NAME).header)
      |                                ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/meta/meta_modelica_data.h:106:35: note: expanded from macro 'MMC_TAGPTR'
  106 | #define MMC_TAGPTR(p)             ((void*)((char*)(p) + 3))
      |                                   ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/util/generic_array.h:48:76: note: expanded from macro 'real_array_copy_data'
   48 | #define real_array_copy_data(src,dst)               simple_array_copy_data(src, &dst, sizeof(modelica_real));
      |                                                                            ^~~
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/util/generic_array.h:46:48: note: passing argument to parameter 'src' here
   46 | void simple_array_copy_data(const base_array_t src, base_array_t* dst, size_t sze);
      |                                                ^
```
By looking at the equation structure with the debugger, there are several expressions with indexing of arrays, such as
for i in 1:10 loop
```modelica
for i in 1:10 loop
    $FUN_3[i] := SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.CollisionIntegral(8, i, diff.Temp / {293.82132, 79.81222, 144.31854, 98.91608, 125.9191, 151.70758, 159.05524, 106.7, 87.28333, 99.7753}[i]);
  end for;
```
or
```modelica
for i in 1:10 loop
    diff.D_H2O_i[i] := 5.959770780848035e-24 * sqrt(55.508435061791985 + 1.0 / {0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948}[i]) * diff.Temp ^ 1.5 * diff.ratioPorTor / (3e6 + time) / {2.641e-10, 2.734e-10, 3.291e-10, 3.1655e-10, 3.1995e-10, 3.542e-10, 3.6595e-10, 3.054e-10, 3.2195e-10, 3.0915e-10}[i] ^ 2.0 / diff.collIntegral_H2O_i[i];
  end for;
```
not sure if they could be problematic for adjacency analysis in more complicated cases than this one, where we prescribe the input variables to be known time-varying expressions. Note that the latter solved equation comes from function inlining, when doing it with the OB, inlining is not applied
```
(assign) diff.D_H2O_i[1] := SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.DiffusivityEffCoeff(diff.Temp, diff.pAnode, 1, 1, 2.641e-10, diff.collIntegral_H2O_i[1], diff.ratioPorTor)
```
I'm not sure if this is the root cause of the failure of the adjacency analysis of the full model, but I guess it would be worth investigating why the generated code of Test is wrong, fix it, and make sure that the obtained results are the same with both backends.

**Comment by casella** [comment]
This is the smallest MWE we could find that reproduces the adjacency analysis failure:
```modelica
model TestPEN
  SOFCPoliMi.Components.FuelCell.PEN pen(iPENmax = 0, vLeak0 = 0)  annotation(
    Placement(transformation(origin = {4, 0}, extent = {{-40, -40}, {40, 40}})));
  Modelica.Blocks.Sources.RealExpression pressure(y = 3e6 + time)  annotation(
    Placement(transformation(origin = {-72, 58}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Blocks.Sources.RealExpression partialPressureAnode[10](y = {0.4 - 0.1*time, 0.1, 0.2, 0.1, 0.15, 0.0, 0.0, 0.05 + 0.1*time, 0.0, 0.0}*(3e6 + time))  annotation(
    Placement(transformation(origin = {-58, -86}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Blocks.Sources.RealExpression partialPressureCathode[10](y = {0.0, 0.0, 0.7 - 0.1*time, 0.0, 0.0, 0.0, 0.0, 0.3 + 0.1*time, 0.0, 0.0}*(3e6 + time)) annotation(
    Placement(transformation(origin = {-30, 80}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Electrical.Analog.Sources.StepCurrent stepCurrent(I = 5)  annotation(
    Placement(transformation(origin = {-76, -4}, extent = {{-10, -10}, {10, 10}}, rotation = 90)));
  Modelica.Electrical.Analog.Basic.Ground ground annotation(
    Placement(transformation(origin = {-78, -62}, extent = {{-10, -10}, {10, 10}})));
equation
  connect(pressure.y, pen.absolutePressureCathode) annotation(
    Line(points = {{-60, 58}, {-32, 58}, {-32, 20}}, color = {0, 0, 127}));
  connect(pressure.y, pen.absolutePressureAnode) annotation(
    Line(points = {{-60, 58}, {-52, 58}, {-52, -26}, {-54, -26}, {-54, -46}, {-32, -46}, {-32, -20}}, color = {0, 0, 127}));
  connect(partialPressureAnode.y, pen.partialPressuresAnode) annotation(
    Line(points = {{-46, -86}, {-16, -86}, {-16, -20}}, color = {0, 0, 127}, thickness = 0.5));
  connect(stepCurrent.p, pen.anodePin) annotation(
    Line(points = {{-76, -14}, {-76, -26}, {-44, -26}, {-44, -8}, {4, -8}}, color = {0, 0, 255}));
  connect(pen.cathodePin, stepCurrent.n) annotation(
    Line(points = {{4, 8}, {-76, 8}, {-76, 6}}, color = {0, 0, 255}));
  connect(partialPressureCathode.y, pen.partialPressuresCathode) annotation(
    Line(points = {{-18, 80}, {0, 80}, {0, 54}, {-16, 54}, {-16, 20}}, color = {0, 0, 127}, thickness = 0.5));
  connect(stepCurrent.p, ground.p) annotation(
    Line(points = {{-76, -14}, {-78, -14}, {-78, -52}}, color = {0, 0, 255}));
annotation(
    uses(Modelica(version = "4.0.0"), SOFCPoliMi),
    __OpenModelica_commandLineOptions = "--newBackend");
end TestPEN;
```
It has 371 equations instead of 16,000 and counting. As for the previous model, you need to install the SOFCPoliMi library first.

**Comment by casella** [comment]
My gut feeling is that `SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.CollisionIntegral` and `SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.DiffusivityEffCoeff` are somehow involved.

**Comment by casella** [comment]
@kabdelhak if you can sort this out, with a bit of luck this model could start working. This would really be a boon for a number of reasons:
- it is a fairly complicated industrial model, not a toy model
- it is fully published, including the source code, so it can be used as a benchmark in the open literature
- although it is not huge, it still has a significant size, about 16,000 equations, which can be arbitrarily increased by taking larger numbers of finite volumes, although going above 100,000 is probably not very meaningful from a modelling point of view; looks like a very nice benchmark for OpenSCALING.
- with the old backend, the current model it takes a couple of minutes to compile, I guess with the new backend this would be reduced to 10-20 seconds max; the advantage would be even larger if we increase the number of volumes
- we still have some initialization issues with this model when using the old backend, see #12941; if we can manage to run it with the NB, @phannebohm knows how to fix them there

Thanks!

**Comment by kabdelhak** [comment]
i found the definition in `SOFCPoliMi.Components.FuelCell.DiffusionLosses.ConcResitanceDGM`:
```
D_O2_i = {Utilities.DiffusivityEffCoeff(Temp, pCathode, 8, i, collDiam_O2_i[i], collIntegral_O2_i[i], ratioPorTor) for i in {1, 2, 3, 4, 5, 6, 7, 9, 10}};
```

It's just written like this so not a weird frontend conversion or something. A possible solution would be to split the resulting for loop into two parts (1-7, 9-10) but this would require some sequence analysis. this is formulated in a way that is meant to be fully scalarized, maybe we can also change the way such things are formulated in the first place to make ranges consistend?

**Comment by phannebohm** [comment]
You could try to automatically convert it to something like this:
```modelica
for $i1 in 1:9 loop
  i = {1, 2, 3, 4, 5, 6, 7, 9, 10}[$i1];
  pen.concLosses.D_O2_i[$i1] = (((5.959770780848035e-24 * sqrt(31.25117191894696 + 1/({0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948})[i]) * pen.concLosses.Temp ^ 1.5 * pen.concLosses.ratioPorTor) / (3e6 + time)) / ({3.054e-10, 3.147e-10, 3.7040000000000003e-10, 3.5785e-10, 3.6125e-10, 3.9550000000000003e-10, 4.2925000000000004e-10, 3.467e-10, 3.6325e-10, 3.5045e-10})[i] ^ 2.0) / pen.concLosses.collIntegral_O2_i[i] ($RES_SIM_144)
end for;
```
i.e. the index list is indexed by the loop iterator. The iterator `$i1` is used on the LHS and the `i` is used on the RHS.

**Comment by kabdelhak** [comment]
> i.e. the index list is indexed by the loop iterator. The iterator `$i1` is used on the LHS and the `i` is used on the RHS.

That's not trivial to handle sadly.

1. the new backend splits up all for equations so that have only one body element, we would need to rewrite everything to actually handle multiple
2. i would need to update everything that tries to evaluate the indices (mainly adjacency) to first gather the top equation and make single replacement rules for all of the tuples (so i still need to scalarize and map the index list to 1:9) to actually evalute the indexed crefs to see what index they point to

**Comment by kabdelhak** [comment]
> i.e. the index list is indexed by the loop iterator. The iterator `$i1` is used on the LHS and the `i` is used on the RHS.

Hmmm but having it as part of the iterator should work, then i will just prepare it once and use it multiple times. Still needs some adjusting because i need to update all iterator functions but that should be very doable

**Comment by casella** [comment]
> It's just written like this so not a weird frontend conversion or something

It is some calculation of chemical-physical properties that depends on most concentrations, but not all. @matteopascali can you comment on that?

**Comment by matteodepascali** [comment]
> > It's just written like this so not a weird frontend conversion or something
> 
> It is some calculation of chemical-physical properties that depends on most concentrations, but not all. [@matteodepascali](https://github.com/matteodepascali) can you comment on that?

Yes, that's it. 
It is the calculation of the diffusivity of species *x*, which depends on all the species of the considered mixture except species *x* itself.

**Comment by kabdelhak** [comment]
Philips approach works when i combine it with the current iterator structure. Currently working on implementing that!

**Comment by kabdelhak** [comment]
The first part should be fixed once #13497 is merged, it runs to another problem with differentiation afterwards, for which i also have a solution, that i am working on now.

**Comment by kabdelhak** [comment]
Somehow my PR closed this, that was not intentional. Some issues still remain.

**Comment by casella** [comment]
Somehow my PR closed this, that was not intentional. Some issues still remain.

It's the AI. If you write "fixes" and then  "#XXXX" in the PR, it just closes issue #xxxx" 😃

**Comment by kabdelhak** [comment]
> Somehow my PR closed this, that was not intentional. Some issues still remain.
> 
> It's the AI. If you write "fixes" and then "#XXXX" in the PR, it just closes issue #xxxx" 😃

Yea Philip told me too :) I guess i have to phrase it differently if it is not fixing it for good.

**Comment by kabdelhak** [comment]
I pushed some more changes getting the model further down the pipeline, i am now stuck at the next issue which is very similar to the one before.

```
  Real[9] pen.concLosses.Xc_in_Rest(max = fill(1.0, 9), min = fill(0.0, 9), unit = fill("1", 9), quantity = fill("MoleFraction", 9)) = cat(1, pen.concLosses.Xc[1:7], pen.concLosses.Xc[9:10]) / pen.concLosses.Xc_Rest;
```

Is a second way you defined this split arrays in the model (`cat(1, pen.concLosses.Xc[1:7], pen.concLosses.Xc[9:10])`) which i need to handle properly. Unfortunately the frontend expands this in the `simplify` step to

```
  Real[9] pen.concLosses.Xc_in_Rest(max = fill(1.0, 9), min = fill(0.0, 9), unit = fill("1", 9), quantity = fill("MoleFraction", 9)) = {pen.concLosses.Xc[1], pen.concLosses.Xc[2], pen.concLosses.Xc[3], pen.concLosses.Xc[4], pen.concLosses.Xc[5], pen.concLosses.Xc[6], pen.concLosses.Xc[7], pen.concLosses.Xc[9], pen.concLosses.Xc[10]} / pen.concLosses.Xc_Rest;
```

@perost Even though this is of course valid and the backend needs to handle it, i would prefer if the frontend did not expand this if `-d=-nfScalarize` or `--newBackend`. Where is this done and where would i need to prevent this expansion?

**Comment by casella** [comment]
> I guess i have to phrase it differently if it is not fixing it for good.

Then you can apply for a new job as ["prompt engineer"](https://aws.amazon.com/what-is/prompt-engineering/). Good grief...

**Comment by casella** [comment]
> i am now stuck at the next issue which is very similar to the one before.

Ugh. Well, I guess the NB should be able to handle that too, since the MLS allows this kind of construct. 

On the other hand, I think we can easily find a workaround for that, basically moving the binding equation to the equation section with a for loop as in the previous case. Let me know if that is necessary to go forward.

**Comment by perost** [comment]
> [@perost](https://github.com/perost) Even though this is of course valid and the backend needs to handle it, i would prefer if the frontend did not expand this if `-d=-nfScalarize` or `--newBackend`. Where is this done and where would i need to prevent this expansion?

It's done in `NFSimplify.simplifyBuiltinCall`. `cat` is always expanded no matter what, but `simplifyBuiltinCall` has an `expand` flag that relies on the `nfScalarize` flag. So you could just skip expansion of `cat` if `expand = true`.

I'm not sure if skipping expansion of all `cat` calls is really a good idea though, it might cause other issues. But I guess you can try it out and see how it works for you.

**Comment by casella** [comment]
Catcalls are a bad practice, so their expansion should be strictly avoided 😁 

Seriously, from what I understand we should skip any expansion of array-involving expressions in the frontend, and leave them to the backend. If the backend can't handle them, it can always scalarize them as a last resort.

@kabdelhak do you agree?

**Comment by kabdelhak** [comment]
> Catcalls are a bad practice, so their expansion should be strictly avoided 😁
> 
> Seriously, from what I understand when --newBackend is selected, we should skip any expansion of array-involving expressions in the frontend, and leave them to the backend. If the backend can't handle them, it can always scalarize them as a last resort.
> 
> [@kabdelhak](https://github.com/kabdelhak) do you agree?

Yes definitely. I am looking into it!

**Comment by kabdelhak** [comment]
> On the other hand, I think we can easily find a workaround for that, basically moving the binding equation to the equation section with a for loop as in the previous case. Let me know if that is necessary to go forward.

For now it would be easier for the new backend if the binding
```
cat(1, pen.concLosses.Xc[1:7], pen.concLosses.Xc[9:10]) / pen.concLosses.Xc_Rest;
```
was instead a for-expression like the others as
```
pen.concLosses.Xc[i] / pen.concLosses.Xc_Rest for i in {1,2,3,4,5,6,7,9,10};
```

But in the long run we want to support the `cat()` function properly since it allows for other constructs that are not doable with the for-expression. Also the for-expression requires a scalarized list, which is also not optimal. Furthermore i need to implement the C-code portion of this new iterator kind, i could not test it yet since the model does not reach runtime.

Unfortunately there is multiple steps missing still to this and like you said, a lot of MSL and other libraries use `cat()` so i want to support it. The final solution for this might even be easier than the scalarized list iterator.

**Comment by casella** [comment]
> The final solution for this might even be easier than the scalarized list iterator.

Good to hear!

**Comment by casella** [comment]
Apologies, in my previous post I actually checked an index-2 problem, so the reported [error message](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarata.err) is actually reasonable; indeed, we have an array of potential state variables, and only the last one will not be a state in the index-reduced model because it is constrained to be constant by the boundary conditions.

Indeed, if you check [SOFCPoliMi.Tests.BenchmarkCammarataIndex1](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarataIndex1.err), it is passing the Causalize steps successfully, so I'd say this issue is resolved 🥳 

If fails immediately afterwards with this error:
```
[/var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/Compiler/NFFrontEnd/NFCeval.mo:1137:9-1137:67:writable] Error: Internal error NFCeval.evalBinaryPow failed to evaluate ‘{0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948} ^ 2.0‘
Error: Internal error NBInitialization.main failed to apply modules!
```
I'm not sure why this happens, I hope it's easy to understand and fix. I'll open another ticket about that.
